File: 프로그래머스/dfs_bfs/여행경로.py
from collections import defaultdict

def dfs(src, dd, answer, length):
    answer.append(src)
    if len(dd[src]) == 0:
        if len(answer) == length:
            return True
        return False

    
    for idx, dest in enumerate(dd[src]):
        del dd[src][idx]
        
        if dfs(dest, dd, answer, length) == True:
            return True
        else:
            answer.pop()
            dd[src].insert(idx, dest)  

def solution(tickets):
    answer = []
    base = "ICN"
    
    dd = defaultdict(list)
    
    for src, dest in tickets:
        dd[src].append(dest)
        dd[src] = sorted(dd[src])
    
    dfs(base, dd, answer, len(tickets)+1)
    return answer
  
# print(solution([["ICN", "JFK"], ["HND", "IAD"], ["JFK", "HND"]]))
print(solution([["ICN", "SFO"], ["ICN", "ATL"], ["SFO", "ATL"], ["ATL", "ICN"], ["ATL","SFO"]]))

File: 프로그래머스/dfs_bfs/여행경로.js
function solution(tickets) {
  const graph = tickets.reduce((acc, [from, to]) => {
    acc[from] ? acc[from].push(to) : (acc[from] = [to]);
    return acc;
  }, {});

  Object.keys(graph).forEach((key) => graph[key].sort());

  function dfs(from, path) {
    path.push(from);
    if (path.length == tickets.length + 1) {
      return path;
    }

    for (let i in graph[from]) {
      const to = graph[from].splice(i, 1)[0];

      if (dfs(to, path)) {
        return path;
      } else {
        path.pop();
        graph[from].splice(i, 0, to);
      }
    }

    return false;
  }

  return dfs("ICN", []);
}

console.log(
  solution([
    ["ICN", "SFO"],
    ["ICN", "ATL"],
    ["SFO", "ATL"],
    ["ATL", "ICN"],
    ["ATL", "SFO"],
  ])
);
